examples/policy_gradient/reinforce/agent.py

```diff
@@ -2,48 +2,185 @@
+  slices. Then, the gradients of the splits are computed by the GPU devices
+  during the execution of train_op(), which finishes by averaging the gradients
+  and updating the shared model weights.
+
+  TODO(ekl) these stages should be combined into one to get better pipelining
+  between SGD minibatches.
+  """
   def __init__(self, name, batchsize, preprocessor, config, use_gpu):
     if not use_gpu:
       os.environ["CUDA_VISIBLE_DEVICES"] = ""
+      devices = ["/cpu:0"]
+    else:
+      devices = config["devices"]
     self.env = BatchedEnv(name, batchsize, preprocessor=preprocessor)
     if preprocessor.shape is None:
       preprocessor.shape = self.env.observation_space.shape
-    self.sess = tf.Session()
-    with tf.name_scope("policy_gradient/train"):
-      with tf.name_scope("proximal_policy_loss"):
-        self.ppo = ProximalPolicyLoss(self.env.observation_space,
-                                      self.env.action_space, preprocessor,
-                                      config, self.sess)
-      with tf.name_scope("adam_optimizer"):
-        self.optimizer = tf.train.AdamOptimizer(config["sgd_stepsize"])
-        self.train_op = self.optimizer.minimize(self.ppo.loss)
-      self.variables = ray.experimental.TensorFlowVariables(self.ppo.loss,
-                                                            self.sess)
-      self.observation_filter = MeanStdFilter(preprocessor.shape, clip=None)
-      self.reward_filter = MeanStdFilter((), clip=5.0)
+    config_proto = tf.ConfigProto(**config["tf_session_args"])
+    self.sess = tf.Session(config=config_proto)
+
+    # Defines the training inputs.
+    self.kl_coeff = tf.placeholder(name="newkl", shape=(), dtype=tf.float32)
+    self.observations = tf.placeholder(tf.float32,
+                                       shape=(None,) + preprocessor.shape)
+    self.advantages = tf.placeholder(tf.float32, shape=(None,))
+
+    action_space = self.env.action_space
+    if isinstance(action_space, gym.spaces.Box):
+      # The first half of the dimensions are the means, the second half are the
+      # standard deviations.
+      self.action_dim = action_space.shape[0]
+      self.logit_dim = 2 * self.action_dim
+      self.actions = tf.placeholder(tf.float32,
+                                    shape=(None, action_space.shape[0]))
+      distribution_class = DiagGaussian
+    elif isinstance(action_space, gym.spaces.Discrete):
+      self.action_dim = action_space.n
+      self.logit_dim = self.action_dim
+      self.actions = tf.placeholder(tf.int64, shape=(None,))
+      distribution_class = Categorical
+    else:
+      raise NotImplemented("action space" + str(type(action_space)) +
+                           "currently not supported")
+    self.prev_logits = tf.placeholder(tf.float32, shape=(None, self.logit_dim))
+
+    # Defines operations for staging input training data.
+    self.num_splits = len(devices)
+    with tf.device("/cpu:0"):
+      stage = StagingArea(
```

Comment:
According to the docs, the staging area memory is located on the device on which it is created, and data copied to the destination device on area.get(). I wonder if it would be more efficient to create a staging area per device so that the copy happens on put() instead.